x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java

```diff
@@ -0,0 +1,116 @@
+ * <p>
+ * The planner tries to push down LIMIT and transform all the unbounded sorts into a TopN.
+ * In some cases it's not possible though, eg.
+ * <p>
+ * from test | sort x | lookup join lookup on x | sort y
+ * <p>
+ * from test | sort x | mv_expand x | sort y
+ * <p>
+ * "sort y" will become a TopN, but "sort x" will remain unbounded, so the query could not be executed.
+ * <p>
+ * In most cases though, following commands can make the previous SORTs redundant,
+ * because it will re-sort previously sorted results (eg. if there is another SORT)
+ * or because the order will be scrambled by another command (eg. a STATS)
+ * <p>
+ * This rule finds and prunes redundant SORTs, attempting to make the plan executable.
+ */
+public class PruneRedundantOrderBy extends OptimizerRules.OptimizerRule<LogicalPlan> {
+
+    @Override
+    protected LogicalPlan rule(LogicalPlan plan) {
+        if (plan instanceof OrderBy || plan instanceof TopN || plan instanceof Aggregate) {
+            IdentityHashMap<OrderBy, Void> redundant = findRedundantSort(((UnaryPlan) plan).child());
+            if (redundant.isEmpty()) {
+                return plan;
+            }
+            return plan.transformUp(p -> {
+                if (redundant.containsKey(p)) {
+                    return ((OrderBy) p).child();
+                }
+                return p;
+            });
+        } else {
+            return plan;
+        }
+    }
+
+    private IdentityHashMap<OrderBy, Void> findRedundantSort(LogicalPlan plan) {
+        List<LogicalPlan> toCheck = new ArrayList<>();
+        toCheck.add(plan);
+
+        IdentityHashMap<OrderBy, Void> result = new IdentityHashMap<>();
+        LogicalPlan p = null;
+        while (true) {
+            if (p == null) {
+                if (toCheck.isEmpty()) {
+                    return result;
+                } else {
+                    p = toCheck.remove(0);
+                }
+            } else if (p instanceof OrderBy ob) {
+                result.put(ob, null);
+                p = ob.child();
+            } else if (p instanceof UnaryPlan unary) {
+                if (unary instanceof Project
+                    || unary instanceof Drop
+                    || unary instanceof Rename
+                    || unary instanceof MvExpand
+                    || unary instanceof Enrich
+                    || unary instanceof RegexExtract
+                    || unary instanceof InlineStats
+                    || unary instanceof Lookup
+                // IMPORTANT
+                // If we introduce window functions or order-sensitive aggs (eg. STREAMSTATS),
+                // the previous sort could actually become relevant
+                // so we have to be careful with plans that could use them, ie. the following
+                    || unary instanceof Filter
+                    || unary instanceof Eval
+                    || unary instanceof Aggregate) {
```

Comment:
This is safe _now_, as long as we don't have window functions (and, in general, functions/aggs that rely on the order of the input).
We can decide to keep it like this for now and review it when we introduce such capabilities, or we can decide to be more paranoid about future regressions and discard such cases, but in this case we won't be able to completely avoid unbounded sorts.